r1cs/gadgets/std/src/groups/nonnative/nonnative_group_gadget.rs

```diff
@@ -0,0 +1,844 @@
+    /// element.
+    pub fn add_unsafe<CS: ConstraintSystem<ConstraintF>>(
+        &self,
+        cs: CS,
+        other: &Self,
+    ) -> Result<Self, SynthesisError> {
+        self.add_internal(cs, other, false)
+    }
+}
+
+
+impl<P, ConstraintF, SimulationF> CondSelectGadget<ConstraintF>
+for GroupAffineNonNativeGadget<P, ConstraintF, SimulationF>
+    where
+        P: SWModelParameters<BaseField = SimulationF>,
+        ConstraintF: PrimeField,
+        SimulationF: PrimeField + SquareRootField
+{
+    #[inline]
+    fn conditionally_select<CS: ConstraintSystem<ConstraintF>>(
+        mut cs: CS,
+        cond: &Boolean,
+        first: &Self,
+        second: &Self,
+    ) -> Result<Self, SynthesisError> {
+        let x = NonNativeFieldGadget::conditionally_select(&mut cs.ns(|| "x"), cond, &first.x, &second.x)?;
+        let y = NonNativeFieldGadget::conditionally_select(&mut cs.ns(|| "y"), cond, &first.y, &second.y)?;
+        let infinity = Boolean::conditionally_select(&mut cs.ns(|| "infinity"), cond, &first.infinity, &second.infinity)?;
+
+        Ok(Self::new(x, y, infinity))
+    }
+
+    fn cost() -> usize {
+        2 * <NonNativeFieldGadget<SimulationF, ConstraintF> as CondSelectGadget<ConstraintF>>::cost() +
+            <Boolean as CondSelectGadget<ConstraintF>>::cost()
+    }
+}
+
+
+
+impl<P, ConstraintF, SimulationF> ConstantGadget<SWProjective<P>, ConstraintF>
+for GroupAffineNonNativeGadget<P, ConstraintF, SimulationF>
+    where
+        P: SWModelParameters<BaseField = SimulationF>,
+        ConstraintF: PrimeField,
+        SimulationF: PrimeField + SquareRootField
+{
+    fn from_value<CS: ConstraintSystem<ConstraintF>>(
+        mut cs: CS,
+        value: &SWProjective<P>,
+    ) -> Self
+    {
+        let value = value.into_affine();
+        let x = NonNativeFieldGadget::from_value(cs.ns(|| "hardcode x"), &value.x);
+        let y = NonNativeFieldGadget::from_value(cs.ns(|| "hardcode y"), &value.y);
+        let infinity = Boolean::constant(value.infinity);
+
+        Self::new(x, y, infinity)
+    }
+
+    fn get_constant(&self) ->SWProjective<P> {
+        let value_proj = SWAffine::<P>::new(
+            self.x.get_value().unwrap(),
+            self.y.get_value().unwrap(),
+            self.infinity.get_value().unwrap()
+        ).into_projective();
+        let x = value_proj.x;
+        let y = value_proj.y;
+        let z = value_proj.z;
+        SWProjective::<P>::new(x, y, z)
+    }
+}
+
+impl<P, ConstraintF, SimulationF> AllocGadget<SWProjective<P>, ConstraintF>
+for GroupAffineNonNativeGadget<P, ConstraintF, SimulationF>
+    where
+        P: SWModelParameters<BaseField = SimulationF>,
+        ConstraintF: PrimeField,
+        SimulationF: PrimeField + SquareRootField
+{
+    #[inline]
+    fn alloc<FN, T, CS: ConstraintSystem<ConstraintF>>(
+        mut cs: CS,
+        value_gen: FN,
+    ) -> Result<Self, SynthesisError>
```

Comment:
It might be useful to do an optimized variant of the alloc for curves with Weierstrass A=0 and small B.
Most of our curves (in particular the secp256k1) are of this form.